osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs

```diff
@@ -107,9 +112,74 @@ public void AddRotation(float angle)
             }
 
             currentRotation += angle;
+
             // rate has to be applied each frame, because it's not guaranteed to be constant throughout playback
             // (see: ModTimeRamp)
-            drawableSpinner.Result.RateAdjustedRotation += (float)(Math.Abs(angle) * (gameplayClock?.GetTrueGameplayRate() ?? Clock.Rate));
+            double time = gameplayClock?.CurrentTime ?? Clock.CurrentTime;
+            double rate = gameplayClock?.GetTrueGameplayRate() ?? Clock.Rate;
+            float rateAdjustedRotation = (float)(angle * rate);
+
+            if (rate > 0)
+            {
+                int direction = Math.Sign(rateAdjustedRotation);
+
+                if (direction != 0 && direction != currentDirection)
+                {
+                    turningPoints.Add((time, currentDelta));
+                    currentDelta = 0;
+                }
+
+                currentDirection = direction;
+            }
+            else if (turningPoints.Count > 0 && time <= turningPoints[^1].time)
+            {
+                currentDelta = turningPoints[^1].delta;
+                currentDirection = Math.Sign(currentDelta);
+                turningPoints.RemoveAt(turningPoints.Count - 1);
+            }
+
+            currentDelta += rateAdjustedRotation;
+            drawableSpinner.Result.RateAdjustedRotation = computeEffectiveRotation();
+        }
+
+        private float computeEffectiveRotation()
```

Comment:
This needs xmldoc to explain what "effective" rotation is.

I guess one of the reasons I'm struggling with this code is that we have...

`currentDelta`
`currentRotation`
`effectiveRotation`
`rateAdjustedRotation`
`lastAngle`

with little documentation as to what each one means. gets very hard to follow.